### PR TITLE
Add option --directory to process tests from one directory

### PR DESCRIPTION
**Pull request by Gormartsen**
Example:

``` bash
 php core/scripts/run-tests.sh --directory core/modules/node
```

It will be needed for contrib tests.

### CODE REVIEW

**Comment by quicksketch** [comment]
Hm, we already have a `--group` option, but I suppose that's not going to cut it because modules could in theory have multiple groups, or it's just a hassle to determine what group name a module decides to use in the first place. So this makes sense to me.

File: core/scripts/run-tests.sh [hunk 1]
```diff
@@ -609,6 +615,31 @@ function simpletest_script_get_test_list() {
...
+      $all_classes = array();
+      foreach ($groups as $group) {
+        $all_classes = array_merge($all_classes, array_keys($group));
+      }
+
+      foreach (file_scan_directory($directory, '/\.test$/') as $file) {
+        $content = file_get_contents($file->uri);
+        preg_match_all('@^class ([^ ]+)@m', $content, $matches);
+        foreach($matches[1] as $test_class) {
+          if(in_array($test_class, $all_classes)) {
```
Comment (quicksketch) [inline core/scripts/run-tests.sh:641]: Missing spaces prior to opening parenthesis here.

```
+        foreach ($matches[1] as $test_class) {
+          if (in_array($test_class, $all_classes)) {
```

File: core/scripts/run-tests.sh [hunk 2]
```diff
@@ -47,6 +47,9 @@
  *              Specify the path and the extension
  *              (i.e. 'core/modules/user/user.test').
  *
+ *
+ *  --directory Run all tests found within the specified file directory.
```
Comment (quicksketch) [inline core/scripts/run-tests.sh]: This should match the `--xml` option where `<path>` is on the same line and then the description is given it's own line.

File: core/scripts/run-tests.sh [hunk 3]
```diff
@@ -276,6 +279,8 @@ All arguments are long options.
   --file      Run tests identified by specific file names, instead of class names.
               Specify the path and the extension
               (i.e. 'core/modules/user/user.test').
+              
+  --directory Run all tests found within the specified file directory.
```
Comment (quicksketch) [inline core/scripts/run-tests.sh]: Same as above, match the formatting of the `--xml` option.

**Comment by Gormartsen** [comment]
> Same as above, match the formatting of the --xml option.

Fixed.

--directory option is mirroring 
https://github.com/drupal/drupal/blob/8.2.x/core/scripts/run-tests.sh#L960

DrupalCI is using it to run tests by contrib modules only.

It is a preparation step to have contrib tests for BackdropCMS.

**Comment by fabianx-ai** [comment]
@quicksketch While --group works for most contrib modules, --directory is more flexible and also easier to use as you can run all tests regardless if you know the test group or not.

**Comment by fabianx-ai** [comment]
@quicksketch That same patch will be also in Drupal 7 soon. I just wait for DrupalCI to not download the patch automatically and fail if its already applied ;).

JFYI

**Comment by Gormartsen** [comment]
Tnx for feedback @LionsAd !
This patch will help me to properly implement contrib tests.

Right now we are running core tests only:(

**Comment by quicksketch** [comment]
Merged into 1.x and 1.4.x for 1.4.4. Thanks @Gormartsen and @LionsAd!

**Pull request merged**